uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -162,31 +153,117 @@ define([
+    function adjustDatesInUI () {
+      var toDate = moment(vm.uiOptions.toDate);
 
-      if (!vm._canCalculateChange()) { return $q.resolve(); }
+      if (vm.uiOptions.toDate) {
+        if (!vm.uiOptions.multipleDays) {
+          if (toDate.isBefore(vm.uiOptions.fromDate)) {
+            vm.uiOptions.toDate = vm.uiOptions.fromDate;
+          }
+        } else {
+          if (toDate.isSameOrBefore(vm.uiOptions.fromDate) || toDate.isAfter(vm.period.end_date)) {
+            vm.uiOptions.toDate = undefined;
 
-      vm.loading.showBalanceChange = true;
+            resetAfterDateChange('to');
+          }
+        }
+      }
+    }
 
-      return vm.request.calculateBalanceChange(vm.selectedAbsenceType.calculation_unit_name)
-        .then(setBalanceChange)
-        .catch(handleError);
+    /**
+     * Calculates and updates opening and closing balances.
+     *
+     * For the opening balance, when in edit mode, if the selected absence type
+     * is the same as the request absence type, the opening balance is the
+     * original opening balance value, otherwise it's the leave balance
+     * remainder.
+     *
+     * The closing balance is the opening balance + change amount.
+     */
+    function calculateOpeningAndClosingBalance () {
+      if (originalOpeningBalance &&
+        originalOpeningBalance.absenceTypeId === vm.selectedAbsenceType.id) {
+        vm.balance.opening = originalOpeningBalance.value || 0;
+      } else {
+        vm.balance.opening = vm.selectedAbsenceType.remainder;
+      }
+      // the change is negative so adding it will actually subtract it
+      vm.balance.closing = vm.balance.opening + vm.balance.change.amount;
     }
 
     /**
      * Change handler when changing no. of days like Multiple Days or Single Day.
-     * It will reset dates, day types, change balance.
+     * It will reset dates and day types and attempt to calculate balance balance change.
+     *
+     * @return {Promise}
      */
     function changeInNoOfDays () {
-      vm._reset();
-      vm._calculateOpeningAndClosingBalance();
+      vm.uiOptions.toDate = vm.uiOptions.fromDate;
+      vm.request.to_date_type = vm.request.from_date_type;
+
+      return $q.resolve()
+        .then(attemptCalculateBalanceChange)
+        .then(vm.changeInNoOfDaysExtended ? vm.changeInNoOfDaysExtended : _.noop);
+    }
+
+    /**
+     * Checks absence period for the changed date,
+     * updates the entitlements if needed
+     * and notifies user if the Absence Period was not found
+     *
+     * @param {String} dateType from|to
+     * @return {Promise}
+     */
+    function checkAbsencePeriod (dateType) {
+      var periodHasBeenChanged;
+
+      return setAbsencePeriod(dateType)
+        .then(function (_periodHasBeenChanged_) {
+          periodHasBeenChanged = _periodHasBeenChanged_;
+
+          return loadCalendar();
+        })
+        .then(function () {
+          return loadAndSetTimeRangesFromWorkPattern(
```

Comment:
should this be done always, or only if the period changed?

Reply:
@AkA84 I think always in order to nicely transit between 'hours' and 'days' absence types.